### PR TITLE
WebsocketHandler — Implementation of Base Class WebSocket Handler to Generically Manage WebSocket Connections.

### PR DESCRIPTION
**Pull request by aidanvillasenor**
### Description
Making a base websocket utility that can be used by specific APIs across monorepo.

### Testing
Tested with one integration test.

### Impact
Pretty isolated

### Other

Fixes #7

### CODE REVIEW

File: src/pysrc/util/websocket.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    def start(self) -> None:
+        loop = asyncio.get_event_loop()
+        self._listener_task = loop.create_task(self._connect_and_listen())
+
+    async def _connect_and_listen(self) -> None:
+        try:
+            print(f"Connecting to WebSocket at {self.base_url}...")
+            self.ws = await asyncio.wait_for(
+                websockets.connect(self.base_url), timeout=10
```
Comment (aidanvillasenor) [inline src/pysrc/util/websocket.py:23]: no retry logic yet. lmk if u want me to add.

File: src/pysrc/test/integration/util/resources/sample_data.pkl
Comment (echavemann) [inline src/pysrc/test/integration/util/resources/sample_data.pkl:1]: ?

File: src/pysrc/test/integration/util/test_websocket.py
Comment (echavemann) [inline src/pysrc/test/integration/util/test_websocket.py:1]: this test doesnt do anything bro

File: src/pysrc/util/websocket.py (hunk 1 above)
Comment (echavemann) [reply src/pysrc/util/websocket.py:23]: add

File: src/pysrc/util/websocket.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                print("Listening for messages from WebSocket...")
+                if self.ws is None:
+                    raise RuntimeError("WebSocket connection is not established.")
+                async for message in self.ws:
+                    try:
+                        data = json.loads(message)
+                        await self.on_message(data)
+                    except json.JSONDecodeError:
+                        print("Failed to parse message")
+                    await asyncio.sleep(0.1)  # small delay for processing, can change
```
Comment (echavemann) [inline src/pysrc/util/websocket.py:50]: this fine, i wouldnt comment that tho

File: src/pysrc/util/websocket.py (hunk 1 above)
Comment (echavemann) [reply src/pysrc/util/websocket.py:23]: can put in on_disconnect

File: src/pysrc/test/integration/util/test_websocket.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import pytest
+import asyncio
+import json
+from websockets.asyncio.server import serve, Server, ServerConnection
+from typing import AsyncGenerator, List
```
Comment (echavemann) [inline src/pysrc/test/integration/util/test_websocket.py:5]: no importing list, next time its 5 bucks for chatgpting ur code

File: poetry.lock [hunk 4]
```diff
@@ -1,4 +1,4 @@
-# This file is automatically @generated by Poetry 1.8.4 and should not be changed by hand.
+# This file is automatically @generated by Poetry 1.8.3 and should not be changed by hand.
```
Comment (echavemann) [inline poetry.lock:1]: outdated poetry pls update

File: src/pysrc/test/integration/util/test_websocket.py
Comment (aidanvillasenor) [reply src/pysrc/test/integration/util/test_websocket.py:1]: lol my fault

File: src/pysrc/util/websocket.py (hunk 1 above)
Comment (aidanvillasenor) [reply src/pysrc/util/websocket.py]: ehhh i dont love. I feel like _connect_and_listen should manage everything related to opening and maintaining the websocket connection, and disconnect everything cleanup related. also seems weird to reconnect every time we disconnect? what if someone calls disconnect on their own when they receieve a certain message bc they actually wanna disconnect? idk j dont love it

File: poetry.lock [hunk 5]
```diff
@@ -237,75 +237,77 @@ files = [
 
 [[package]]
 name = "numpy"
-version = "2.1.2"
```
Comment (aidanvillasenor) [inline poetry.lock:240]: what do I change and not change in this file? is there a rule of thumb?

Comment (aidanvillasenor) [reply poetry.lock:240]: seems strange that my code is changing numpy lol

File: src/pysrc/util/websocket.py (hunk 1 above)
Comment (echavemann) [reply src/pysrc/util/websocket.py]: i am 100000d that we want iti in on_disconnect but we can find out the hard way too

File: src/pysrc/test/integration/util/resources/sample_data.pkl
Comment (echavemann) [inline src/pysrc/test/integration/util/resources/sample_data.pkl:1]: this is back again? dude?

File: src/pysrc/util/websocket.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+from websockets.asyncio.client import ClientConnection
+from typing import Optional
+
+
+class WebSocketClient(ABC):
+    def __init__(
+        self, base_url: str, max_retries: int = 5, retry_delay: int = 5
+    ) -> None:
+        self.base_url = base_url
+        self.max_retries = max_retries  # -1 for infinite
```
Comment (echavemann) [inline src/pysrc/util/websocket.py:14]: make it optional, this isnt c, we dont need a sentinel value

File: src/pysrc/util/websocket.py (hunk 2 above)
Comment (echavemann) [reply src/pysrc/util/websocket.py:67]: not resolved it would appear

File: poetry.lock (hunk 5 above)
Comment (echavemann) [reply poetry.lock:240]: this change is fine, as long as the poetry versions match

**Review by echavemann — APPROVED** [review]

File: src/pysrc/util/websocket.py (hunk 1 above)
Comment (stevenewald) [reply src/pysrc/util/websocket.py]: 1. this should be a state machine
2. if it isn't a state machine (reasonable if we just wanna get something working), i think i agree with aidan. if on_disconnect should reconnect, it doesn't make sense for it to have the retry logic

Comment (echavemann) [reply src/pysrc/util/websocket.py]: Let's find out

Comment (stevenewald) [reply src/pysrc/util/websocket.py]: @echavemann i am genuinely curious ur reasoning

**Pull request merged**